apps/api-extractor/src/generators/ApiModelGenerator.ts

```diff
@@ -831,11 +835,13 @@ export class ApiModelGenerator {
       const releaseTag: ReleaseTag = apiItemMetadata.effectiveReleaseTag;
       const isOptional: boolean =
         (astDeclaration.astSymbol.followedSymbol.flags & ts.SymbolFlags.Optional) !== 0;
+      const isProtected: boolean = (astDeclaration.modifierFlags & ts.ModifierFlags.Protected) !== 0;
```

Comment:
I thought for a bit about whether this should be an `enum` (or even bitfield like `ts.ModifierFlags`) .  TypeScript's official access modifiers are `public`, `protected`, and `private` (although that they could always introduce more in the future).  And although `private` is not accessible by API consumers, it does impact them because the compiler forbids these names from being reused in subclasses (which is why they get emitted into the .d.ts files).  API Extractor has considered possibly introducing other access modifiers such as `@friend` (e.g. for members that are `public` so unit tests can access them, but not meant to be used by other modules).

But I ended up agreeing with your `boolean` design. 👍  Specifically I'm thinking that it's more natural for `private`, `friend`, `sealed`, `virtual`, `override`, etc all to be simple boolean "modifiers", without trying to model their mutual exclusivity in the type system.

Reply:
> And although `private` is not accessible by API consumers, it does impact them because the compiler forbids these names from being reused in subclasses (which is why they get emitted into the .d.ts files).

Ah... I was wondering why they're emitted to `.d.ts`. That makes sense.

> But I ended up agreeing with your `boolean` design.

SG, plus we can always iterate when building support for additional visibility modifiers.

Reply:
> Specifically I'm thinking that it's more natural for `private`, `friend`, `sealed`, `virtual`, `override`, etc all to be simple boolean "modifiers", without trying to model their mutual exclusivity in the type system.

@zelliott  I was thinking about this some more:  If we do plan to eventually support  `private`, `friend`, `sealed`, `virtual`, `override`, etc, it might be clunky to introduce a dedicated mixin class for each one. 

🤔 The idea behind mixin classes was to prevent impossible combinations: For example, a TypeScript constructor can accept function parameters, but NOT a function return value.  If the model allowed an `ApiConstructor.returnTypeExcerpt` field, this could be potentially confusing:  A person who is implementing a documentation tool would see that and wonder whether constructors can have return values (TypeScript is tricky!).  A wise person would be paranoid about this; they would accommodate it in their documentation template, which is wasteful extra effort.

This led to lots of very narrow mixins, e.g. `ApiOptionalMixin` and `ApiStaticMixin` and so forth. At one point I made a big Excel matrix of features <-> API Items, hoping to combine identical feature vectors into common classes. It was surprising to find that almost every vector was different. (For example, a constructor can be `protected` but not `virtual`.) I think this orthogonality isn't accidental, but actually an indirect consequence of good language design.

So if we want to avoid too many mixins, we would do it by allowing impossible combinations: For example we might say that `static` gets `ApiStaticMixin` because it's an "important" aspect (e.g. `static` alters your declaration reference), whereas maybe `protected` and `virtual` are unimportant "details" and thus could be stored in `ApiItem.details.isProtected` or perhaps a `ApiMemberMixin.isProtected`. This important-vs-unimportant criterion would tolerate the incorrect `ApiConstructor.isProtected` field to simplify the API.

Another idea would be to embrace the current model, and make peace with having tons of mixins. An upside is that we avoid having to explain a criterion to end users, and by being accurate `api-extractor-model` might teach something about the TypeScript language.  If the extensive inheritance becomes a performance concern, perhaps we could consider an optimization where everything is rolled into one monolithic class internally, but then the public API rely on `interface` types to hide the impossible members. 

I don't know the right answer. Let me know if you have any thoughts. 😋

Reply:
I don't know the right answer either, but I'll share my perspective. As someone who's new to API Extractor and many of TypeScript's nuances and advanced language features, I've appreciated the accuracy of the generated model. The benefit you mention has resonated with me in my past few weeks working closely with API Extractor. I agree that it would be confusing if `ApiConstructor.returnTypeExcerpt` existed, just as it would be confusing if `ApiConstructor.isReadonly` existed. I haven't found the large number of mixins particularly troublesome to work with (yet), and I'd prioritize clarity in the generated model over performance at this time (but of course we should continue to watch this over time).

Reply:
Ok cool. Then I'm good with `ApiProtectedMixin` as-is. 👍